src/python/pants/backend/go/goals/generate.py

```diff
@@ -0,0 +1,285 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+import os
+import re
+import shlex
+import string
+from dataclasses import dataclass
+from typing import Mapping
+
+from pants.backend.go.target_types import GoPackageSourcesField
+from pants.backend.go.util_rules import first_party_pkg, goroot, sdk
+from pants.backend.go.util_rules.first_party_pkg import (
+    FallibleFirstPartyPkgAnalysis,
+    FallibleFirstPartyPkgDigest,
+    FirstPartyPkgAnalysis,
+    FirstPartyPkgAnalysisRequest,
+    FirstPartyPkgDigestRequest,
+)
+from pants.backend.go.util_rules.goroot import GoRoot
+from pants.build_graph.address import Address
+from pants.engine.env_vars import EnvironmentVars, EnvironmentVarsRequest
+from pants.engine.fs import DigestContents, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.native_engine import EMPTY_DIGEST, AddPrefix, Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import collect_rules, goal_rule, rule, rule_helper
+from pants.engine.target import Targets
+from pants.option.option_types import StrListOption
+from pants.util.strutil import softwrap
+
+logger = logging.getLogger(__name__)
+
+
+# Adapted from Go toolchain.
+# See https://github.com/golang/go/blob/master/src/cmd/go/internal/generate/generate.go and
+# https://github.com/golang/go/blob/cc1b20e8adf83865a1dbffa259c7a04ef0699b43/src/os/env.go#L16-L96
+#
+# Original copyright:
+#   // Copyright 2011 The Go Authors. All rights reserved.
+#   // Use of this source code is governed by a BSD-style
+#   // license that can be found in the LICENSE file.
+
+
+_GENERATE_DIRECTIVE_RE = re.compile(rb"^//go:generate[ \t](.*)$")
+
+
+class GoGenerateGoalSubsystem(GoalSubsystem):
+    name = "go-generate"
+    help = softwrap(
+        """
+        Run each command in a package described by a `//go:generate` directive. This is equivalent to running
+        `go generate` on a Go package.
+
+        Note: Just like with `go generate`, the `go-generate` goal is never run as part of the build and
+        must be run manually to invoke the commands described by the //go:generate directives.
+
+        See https://go.dev/blog/generate for details.
+        """
+    )
+
+    env_vars = StrListOption(
+        default=["LANG", "LC_CTYPE", "LC_ALL", "PATH"],
+        help=softwrap(
+            """
+            Environment variables to set when invoking generator programs.
+            Entries are either strings in the form `ENV_VAR=value` to set an explicit value;
+            or just `ENV_VAR` to copy the value from Pants's own environment.
+            """
+        ),
+        advanced=True,
+    )
```

Comment:
You'll save @chrisjrn some work if this uses the new Environment Aware subsystems from https://github.com/pantsbuild/pants/pull/16965

(ack this is new and my code review is delayed)

Reply:
Will look into it.